File: Source/DogFight/Private/GameService/DamageCalculatorService.cpp
#include "GameService/DamageCalculatorService.h"

#include "DamageCalculator/DamageCalculatorBase.h"
#include "DamageType/ExtendedDamageInstance.h"

UDamageCalculatorBase* UDamageCalculatorService::CreateDamageCalculator(const FString& ScriptPath, UObject* Instigator)
{
	// Create calculator first
	auto NewCalculator = LoadDamageCalculator(ScriptPath, Instigator);
	if (!NewCalculator)
	{
		return nullptr;
	}

	InitializeDamageCalculator(NewCalculator);
	return NewCalculator;
}

void UDamageCalculatorService::InitializeDamageCalculator(UDamageCalculatorBase* InCalculator)
{
	if (!InCalculator)
	{
		return;
	}

	auto DamageInstanceList = InCalculator->GetDamageInstanceList();
	for (auto InstanceName : DamageInstanceList)
	{
		if (InstanceName.IsEmpty())
			continue;

		auto NewInstance = LoadDamageInstance(InstanceName, InCalculator);
		if (NewInstance)
		{
			InCalculator->RegisterNewDamageInstance(FName(InstanceName), NewInstance);
		}
	}
}

UClass* UDamageCalculatorService::GetDamageCalculatorClass()
{
	return UDamageCalculatorBase::StaticClass();
}

UClass* UDamageCalculatorService::GetDamageInstanceClass()
{
	return UExtendedDamageInstance::StaticClass();
}

File: Content/Script/DogFight/Services/DamageCalculatorService.lua
require "UnLua"

local DamageCalculatorService = Class("DogFight.Services.GameServiceBase")

function DamageCalculatorService:LoadDamageCalculator(InPath, Instigator)
    local DefaultClass = self:GetDamageCalculatorClass()
    if DefaultClass == nil then
        return
    end

    local NewInstance = NewObject(DefaultClass, Instigator, nil, self:GetCalculatorScriptPath(InPath))
    return NewInstance
end

function DamageCalculatorService:LoadDamageInstance(InPath, Instigator)
    local DefaultClass = self:GetDamageInstanceClass()
    if DefaultClass == nil then
        return
    end

    local NewInstance = NewObject(DefaultClass, Instigator, nil, self:GetInstanceScriptPath(InPath))
    return NewInstance
end

function DamageCalculatorService:GetCalculatorScriptPath(InPath)
    return "DogFight.DamageCalculator." .. InPath
end

function DamageCalculatorService:GetInstanceScriptPath(InPath)
    return "DogFight.DamageInstance." .. InPath
end

return DamageCalculatorService
